### PR TITLE
8343449: Nashorn method handle debug logging breaks with log4j-jul

### PR DESCRIPTION
**Pull request by ppkarwasz**
The `MethodHandleFactory` and especially its `FUNC` static field are heavily used during the static initialization process.

To prevent nullability problems in the static `MethodHandleFactory` fields, `StandardMethodHandleFunctionality` relies on `DebugLogger.DISABLED_LOGGER.isEnabled()` being `false` during the initialization process.
If this assumption is true, static `MethodHandleFactory` fields will not be accessed by `StandardMethodHandleFunctionality`.
If this assumption, however, fails (as described in apache/logging-log4j2#3119) the static initialization fails with some unforeseen NPEs.

We improve the above machanism by:

- Replacing the `StandardMethodHandleFunctionality.log` field with a static field in the top level class.
- The new `log` field is placed last, so it holds the value `null` until the end of static initialization.
- All the usages of the `log` fields are checked to ensure that they contain a null-check.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace

### Issue
 * [JDK-8343449](https://bugs.openjdk.org/browse/JDK-8343449): Nashorn method handle debug logging breaks with log4j-jul (**Bug** - P4)

### Reviewers
 * [Attila Szegedi](https://openjdk.org/census#attila) (@szegedi - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/nashorn.git pull/19/head:pull/19` \
`$ git checkout pull/19`

Update a local copy of the PR: \
`$ git checkout pull/19` \
`$ git pull https://git.openjdk.org/nashorn.git pull/19/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19`

View PR using the GUI difftool: \
`$ git pr show -t 19`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/nashorn/pull/19.diff">https://git.openjdk.org/nashorn/pull/19.diff</a>

</details>

### CODE REVIEW

File: src/org.openjdk.nashorn/share/classes/org/openjdk/nashorn/internal/lookup/MethodHandleFactory.java [hunk 1]
```diff
@@ -369,13 +369,11 @@ protected static String describe(final Object... data) {
         }
 
         public MethodHandle debug(final MethodHandle master, final String str, final Object... args) {
-            if (log.isEnabled()) {
-                if (PRINT_STACKTRACE) {
-                    stacktrace(log);
-                }
-                return addDebugPrintout(log, Level.INFO, master, Integer.MAX_VALUE, false, str + ' ' + describe(args));
+            // The static `log` field can be `null` if the `MethodHandlerFactory` class has not been initialized yet.
```
Comment (szegedi) [inline src/org.openjdk.nashorn/share/classes/org/openjdk/nashorn/internal/lookup/MethodHandleFactory.java:372]: Don't you need the null check here then too?

Comment (ppkarwasz) [reply src/org.openjdk.nashorn/share/classes/org/openjdk/nashorn/internal/lookup/MethodHandleFactory.java:372]: Unless I missed some code path, the `addDebugPrintout` and `stacktrace` both handle a `null` input correctly.

https://github.com/openjdk/nashorn/blob/7bd13f5684711f60d838229f594874535336f370/src/org.openjdk.nashorn/share/classes/org/openjdk/nashorn/internal/lookup/MethodHandleFactory.java#L282-L287

https://github.com/openjdk/nashorn/blob/7bd13f5684711f60d838229f594874535336f370/src/org.openjdk.nashorn/share/classes/org/openjdk/nashorn/internal/lookup/MethodHandleFactory.java#L194-L204

**Comment by szegedi** [comment]
Okay, I dug into this a bit more and also in the analysis in https://github.com/apache/logging-log4j2/issues/3119. Why would `DebugLogger.DISABLED_LOGGER.isEnabled()` return true when used with `log4j-jul`? `DISABLED_LOGGER` is created with `Level.OFF`, so it's always disabled (as far as `DebugLogger.isEnabled` is concerned - it's a final field set by `DebugLogger` ctor if the level is `OFF`.)


OTOH, it's _possible_ that maybe `logger.isLoggable(level)` check on `MethodHandleFactory.java:281`  might return true now? In the stack trace in https://github.com/apache/logging-log4j2/issues/3119 it is invoked from `MethodHandleFactory.java:376` which passed `Level.INFO`, so is it the case that a logger created with `Level.OFF` returned true from `isLoggable(Level.INFO)` call? Is that expected behavior in `log4j-jul`?

**Comment by ppkarwasz** [comment]
> Okay, I dug into this a bit more and also in the analysis in [apache/logging-log4j2#3119](https://github.com/apache/logging-log4j2/issues/3119). Why would `DebugLogger.DISABLED_LOGGER.isEnabled()` return true when used with `log4j-jul`? `DISABLED_LOGGER` is created with `Level.OFF`, so it's always disabled (as far as `DebugLogger.isEnabled` is concerned - it's a final field set by `DebugLogger` ctor if the level is `OFF`.)

The comparison between the configured level and `Level.OFF` is not direct and goes through JUL. The code:

https://github.com/openjdk/nashorn/blob/2eb88e4024023ee8e9baacb7736f914e3aa68aa4/src/org.openjdk.nashorn/share/classes/org/openjdk/nashorn/internal/runtime/logging/DebugLogger.java#L76-L81

is basically equivalent to:

```java
this.logger.setLevel(loggerLevel);
this.enabled = this.logger.getLevel() != Level.OFF;
```

In `log4j-jul` the `getLevel()` method always returns `null`. We will change that in apache/logging-log4j2#3125, but this is more of a workaround that a real fix.

> OTOH, it's _possible_ that maybe `logger.isLoggable(level)` check on `MethodHandleFactory.java:281` might return true now? In the stack trace in [apache/logging-log4j2#3119](https://github.com/apache/logging-log4j2/issues/3119) it is invoked from `MethodHandleFactory.java:376` which passed `Level.INFO`, so is it the case that a logger created with `Level.OFF` returned true from `isLoggable(Level.INFO)` call? Is that expected behavior in `log4j-jul`?

A logger can not be created with `Level.OFF`, because Log4j API, pretty much like SLF4J, does not have an equivalent of `setLevel()`. Calling `Logger.setLevel()` is a no-op.

If Log4j API is used together with the Log4j Core implementation, we have an option (cf. [`log4j2.julLoggerAdapter`](https://logging.apache.org/log4j/2.x/log4j-jul.html#log4j2.julLoggerAdapter)) that allows users to modify the configuration of Log4j Core through `j.u.l.Logger`. We disabled that option in `2.24.0`, because it was causing problems to users (see apache/logging-log4j2#2353).

In general the usage of `Logger.setLogger()` is problematic, because it is hard to distinguish between:

- Legitimate usages: a class should be able to modify the configuration of its "own" logger.
- Abusive usages: some legacy libraries just set the level of the root logger to whatever they want and break users configurations.

Therefore I believe that the code that modifies the logging configuration should be in `org.openjdk.nashorn.shell`, which is a standalone application and should be able to control its logging configuration. The reusable `org.openjdk.nashorn` library should probably abstain from modifying the logging configuration.

File: src/org.openjdk.nashorn/share/classes/org/openjdk/nashorn/internal/lookup/MethodHandleFactory.java [hunk 2]
```diff
@@ -369,13 +369,11 @@ protected static String describe(final Object... data) {
...
+            if (PRINT_STACKTRACE) {
```
Comment (szegedi) [inline src/org.openjdk.nashorn/share/classes/org/openjdk/nashorn/internal/lookup/MethodHandleFactory.java:373]: `stacktrace` itself checks `PRINT_STACKTRACE` so this `if` could also be removed.

Comment (ppkarwasz) [reply src/org.openjdk.nashorn/share/classes/org/openjdk/nashorn/internal/lookup/MethodHandleFactory.java]: Fixed in https://github.com/openjdk/nashorn/pull/19/commits/ee8edb3f800c9f09753a25d53e6b05633ca5e62d

**Comment by szegedi** [comment]
I think the initialization order here is rather brittle and unfortunate. This could all be avoided if `TRACE`, `TRACE_RETURN`, and `TRACE_RETURN_VOID` were looked up without going through `FUNC` which in this case is not really doing anything beyond exception wrapping. Basically, I'd move them before the `FUNC` declaration as
```java
    private static final MethodHandle TRACE;
    private static final MethodHandle TRACE_RETURN;
    private static final MethodHandle TRACE_RETURN_VOID;
    static {
        try {
            TRACE = LOOKUP.findStatic(MethodHandleFactory.class, "traceArgs",   MethodType.methodType(void.class, DebugLogger.class, String.class, int.class, Object[].class));
            TRACE_RETURN = LOOKUP.findStatic(MethodHandleFactory.class, "traceReturn", MethodType.methodType(Object.class, DebugLogger.class, Object.class));
            TRACE_RETURN_VOID = LOOKUP.findStatic(MethodHandleFactory.class, "traceReturnVoid", MethodType.methodType(void.class, DebugLogger.class));
        } catch (NoSuchMethodException | IllegalAccessException e) {
            throw new LookupException(e);
        }
    }
```
And also keeping the rest of your changes.

**Comment by szegedi** [comment]
Alright, I created a JBS issue for this. If you sign the OCA we can move forward with this.

**Comment by ppkarwasz** [comment]
> Alright, I created a JBS issue for this. 

Thanks.

> If you sign the OCA we can move forward with this.

I submitted it on Tuesday, but it still has a status of "Under review".

**Comment by szegedi** [comment]
> I submitted it on Tuesday, but it still has a status of "Under review".

That's fine, I'm sure it'll work its way through in a few days.

File: src/org.openjdk.nashorn/share/classes/org/openjdk/nashorn/internal/lookup/MethodHandleFactory.java [hunk 3]
```diff
@@ -111,12 +124,14 @@ public static MethodHandleFunctionality getFunctionality() {
...
-    private static final MethodHandle TRACE_RETURN_VOID = FUNC.findStatic(LOOKUP, MethodHandleFactory.class, "traceReturnVoid", MethodType.methodType(void.class, DebugLogger.class));
-
     private static final String VOID_TAG = "[VOID]";
 
+    // A lot of static fields use FUNC for lookups.
+    //
+    // We keep the logger as the last static field in this class, to ensure that no logging is attempted, until
+    // the rest of the static fields are initialized.
+    private static final DebugLogger log = Context.getContext().getLogger(StandardMethodHandleFunctionality.class);
+
```
[the remark this answers is not in the document]
Comment (ppkarwasz) [reply src/org.openjdk.nashorn/share/classes/org/openjdk/nashorn/internal/lookup/MethodHandleFactory.java]: I moved it to the top of the class in https://github.com/openjdk/nashorn/pull/19/commits/bec8c8bebccd31a0c458758fc2a725f4231a8b27

File: src/org.openjdk.nashorn/share/classes/org/openjdk/nashorn/internal/lookup/MethodHandleFactory.java [hunk 4]
```diff
@@ -274,14 +289,14 @@ public static MethodHandle addDebugPrintout(final MethodHandle mh, final int par
...
-
         //if there is no logger, or if it's set to log only coarser events
         //than the trace level, skip and return
         if (logger == null || !logger.isLoggable(level)) {
             return mh;
         }
 
+        final MethodType type = mh.type();
+
         assert TRACE != null;
```
[the remark this answers is not in the document]
Comment (ppkarwasz) [reply src/org.openjdk.nashorn/share/classes/org/openjdk/nashorn/internal/lookup/MethodHandleFactory.java]: Fixed in https://github.com/openjdk/nashorn/pull/19/commits/bec8c8bebccd31a0c458758fc2a725f4231a8b27

File: src/org.openjdk.nashorn/share/classes/org/openjdk/nashorn/internal/lookup/MethodHandleFactory.java [hunk 5]
```diff
@@ -369,13 +378,10 @@ protected static String describe(final Object... data) {
...
         public MethodHandle debug(final MethodHandle master, final String str, final Object... args) {
-            if (log.isEnabled()) {
-                if (PRINT_STACKTRACE) {
-                    stacktrace(log);
-                }
-                return addDebugPrintout(log, Level.INFO, master, Integer.MAX_VALUE, false, str + ' ' + describe(args));
-            }
-            return master;
+            // The static `log` field can be `null` if the `MethodHandlerFactory` class has not been initialized yet.
+            // Both these methods are null-safe
```
[the remark this answers is not in the document]
Comment (ppkarwasz) [reply src/org.openjdk.nashorn/share/classes/org/openjdk/nashorn/internal/lookup/MethodHandleFactory.java]: Removed in https://github.com/openjdk/nashorn/pull/19/commits/bec8c8bebccd31a0c458758fc2a725f4231a8b27

File: src/org.openjdk.nashorn/share/classes/org/openjdk/nashorn/internal/lookup/MethodHandleFactory.java [hunk 6]
```diff
@@ -569,7 +575,9 @@ public MethodHandle findVirtual(final MethodHandles.Lookup explicitLookup, final
         @Override
         public SwitchPoint createSwitchPoint() {
             final SwitchPoint sp = new SwitchPoint();
-            log.log(TRACE_LEVEL, "createSwitchPoint ", sp);
+            if (log != null) {
```
[the remark this answers is not in the document]
Comment (ppkarwasz) [reply src/org.openjdk.nashorn/share/classes/org/openjdk/nashorn/internal/lookup/MethodHandleFactory.java]: Fixed in https://github.com/openjdk/nashorn/pull/19/commits/bec8c8bebccd31a0c458758fc2a725f4231a8b27

File: src/org.openjdk.nashorn/share/classes/org/openjdk/nashorn/internal/lookup/MethodHandleFactory.java [hunk 7]
```diff
@@ -582,7 +590,9 @@ public MethodHandle guardWithTest(final SwitchPoint sp, final MethodHandle befor
         @Override
         public MethodType type(final Class<?> returnType, final Class<?>... paramTypes) {
             final MethodType mt = MethodType.methodType(returnType, paramTypes);
-            log.log(TRACE_LEVEL, "methodType ", returnType, " ", Arrays.toString(paramTypes), " ", mt);
+            if (log != null) {
```
[the remark this answers is not in the document]
Comment (ppkarwasz) [reply src/org.openjdk.nashorn/share/classes/org/openjdk/nashorn/internal/lookup/MethodHandleFactory.java]: Fixed in https://github.com/openjdk/nashorn/pull/19/commits/bec8c8bebccd31a0c458758fc2a725f4231a8b27

**Comment by szegedi** [comment]
Hi @ppkarwasz! I just released Nashorn 15.5 including this change.

**Pull request closed without merging**